pkg/image/registryclient/alternativesources.go

```diff
@@ -0,0 +1,193 @@
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
+			}
+			for _, m := range rdm.Mirrors {
+				if sourceMatches {
+					klog.V(2).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+					mRef, err := imagereference.Parse(m)
+					if err != nil {
+						return nil, err
+					}
+					imageSources = append(imageSources, mRef)
+				}
+			}
+		}
+	}
+	uniqueMirrors := make([]imagereference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[imagereference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	if len(imageSources) == 0 {
+		imageSources = append(imageSources, imageRef.AsRepository())
+		return imageSources, nil
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+// PreferredImageSource chooses appropriate DockerImageReference for a Context from possible image sources gathered from
+// ImageContentSourcePolicy objects and user-passed image. PreferredImageSource will lookup from cluster
+// if alwaysTryAlternativeSources is true, otherwise will lookup from file if file given, or from
+// original image-reference of user-given image (that may be different from user-given in case of mirrored images).
+func (c *Context) PreferredImageSource(image imagereference.DockerImageReference, alwaysTryAlternativeSources, insecure bool) (imagereference.DockerImageReference, distribution.Repository, distribution.ManifestService, error) {
+	var (
+		repo          distribution.Repository
+		replacedImage imagereference.DockerImageReference
+		manifests     distribution.ManifestService
+		err           error
+	)
+	ctx := context.TODO()
+
+	// alwaysTryAlternativeSources is false when don't want implicit alternative image source lookup
+	// ICSPFile indicates explicit alternative image source lookup
+	if len(c.ImageSources.ICSPFile) == 0 && !alwaysTryAlternativeSources {
+		if repo, err = c.Repository(ctx, image.DockerClientDefaults().RegistryURL(), image.RepositoryName(), insecure); err != nil {
+			return imagereference.DockerImageReference{}, nil, nil, err
+		}
+		manifests, err = repo.Manifests(ctx)
+		return image, repo, manifests, err
+	}
+
+	var altSources []imagereference.DockerImageReference
+	// error if given an icsp file and don't successfully connect to image repository
+	if len(c.ImageSources.ICSPFile) > 0 {
+		if err = c.addImageSourcePoliciesFromFile(); err != nil {
```

Comment:
With the above comments about nil checks for `ICSPFile` and `ICSPClientFn` you'd just call here `addImageSourcePoliciesFromFile` and `addICSPsFromCluster` and have a single for loop over the alternative sources. WDYT?

Reply:
looking - after pushing the last commit i felt something was still not quite right, I'll update w/ your suggestion :)

Reply:
yes, much better, thanks for seeing that :blush: